instrumentation/opentelemetry-instrumentation-kafka-python/src/opentelemetry/instrumentation/kafka/utils.py

```diff
@@ -0,0 +1,260 @@
+            del self.tokens[consumer]
+
+    def close(self, kafka_consumer: KafkaConsumer = None):
+        if kafka_consumer:
+            self.close_span(kafka_consumer)
+            self.detach_context(kafka_consumer)
+        else:
+            for consumer in self.spans:
+                self.close_span(consumer)
+            for consumer in self.tokens:
+                self.detach_context(consumer)
+
+
+class KafkaPropertiesExtractor:
+    @staticmethod
+    def extract_bootstrap_servers(instance):
+        return instance.config.get("bootstrap_servers")
+
+    @staticmethod
+    def _extract_argument(key, position, default_value, args, kwargs):
+        if len(args) > position:
+            return args[position]
+        return kwargs.get(key, default_value)
+
+    @staticmethod
+    def extract_send_topic(args):
+        """extract topic from `send` method arguments in KafkaProducer class"""
+        if len(args) > 0:
+            return args[0]
+        return "unknown"
+
+    @staticmethod
+    def extract_send_value(args, kwargs):
+        """extract value from `send` method arguments in KafkaProducer class"""
+        return KafkaPropertiesExtractor._extract_argument(
+            "value", 1, None, args, kwargs
+        )
+
+    @staticmethod
+    def extract_send_key(args, kwargs):
+        """extract key from `send` method arguments in KafkaProducer class"""
+        return KafkaPropertiesExtractor._extract_argument(
+            "key", 2, None, args, kwargs
+        )
+
+    @staticmethod
+    def extract_send_headers(args, kwargs):
+        """extract headers from `send` method arguments in KafkaProducer class"""
+        return KafkaPropertiesExtractor._extract_argument(
+            "headers", 3, None, args, kwargs
+        )
+
+    @staticmethod
+    def extract_send_partition(instance, args, kwargs):
+        """extract partition `send` method arguments, using the `_partition` method in KafkaProducer class"""
+        topic = KafkaPropertiesExtractor.extract_send_topic(args)
+        key = KafkaPropertiesExtractor.extract_send_key(args, kwargs)
+        value = KafkaPropertiesExtractor.extract_send_value(args, kwargs)
+        partition = KafkaPropertiesExtractor._extract_argument(
+            "partition", 4, None, args, kwargs
+        )
+        key_bytes = instance._serialize(
+            instance.config["key_serializer"], topic, key
+        )
+        value_bytes = instance._serialize(
+            instance.config["value_serializer"], topic, value
+        )
+        valid_types = (bytes, bytearray, memoryview, type(None))
+        if (
+            type(key_bytes) not in valid_types
+            or type(value_bytes) not in valid_types
+        ):
+            return None
+        return instance._partition(
+            topic, partition, key, value, key_bytes, value_bytes
+        )
+
+
+HookT = Callable[[Span, List, Dict], None]
+
+
+def dummy_callback(span, args, kwargs):
```

Comment:
Is there a need for this callback? Usually we just do a `if callable(...)` check.

Reply:
this was my original approach, but a previous reviewer suggested using this approach
I am ok with both, just want to be aligned with the way we do it in other instrumentations

Reply:
Oh interesting. Any strong reason from the previous review to do it this way? If I recall, most of our instrumentations simply take the [value](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L286) as is and perform either [None](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L162) checks or `callables`. Of course, `callable` is preferred.